Add _WKWebExtensionControllerConfiguration to control extension storage.

xeenon opened the pull request:
#### 3bd3d318889c814347c566a45b349c96f078720b
<pre>
Add _WKWebExtensionControllerConfiguration to control extension storage.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248665">https://bugs.webkit.org/show_bug.cgi?id=248665</a>

Reviewed by Brian Weinstein.

_WKWebExtensionControllerConfiguration allows for default persistent storage, non-persistent,
and identifier based storage. With this, we can now store state for extension contexts and have
a directory structure to store extension data in the future (like local storage, etc).

Default storage is placed at: ~/Library/WebKit/&lt;app-id&gt;/WebExtensions/Default/&lt;extension-id&gt;
Non-persistent storage is placed at: ~/Library/WebKit/&lt;app-id&gt;/WebExtensions/&lt;identifier&gt;/&lt;extension-id&gt;
Non-persistent storage is not saved to disk.

Only extension contexts that have a custom uniqueIdentifier will be persistent as well, since the default
uniqueIdentifier is a random UUID, persisting those would cause state to be saved under random IDs that
can&apos;t be reused later.

* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.h: Removed UUID helpers since UUID has Cocoa conversions now.
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm: Ditto.
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration copyWithZone:]): Adjust access to _webExtensionController now that it is RetainPtr.
(-[WKWebViewConfiguration _strongWebExtensionController]): Ditto.
(-[WKWebViewConfiguration _webExtensionController]): Don&apos;t use a LazyInitialized since we don&apos;t require a
webExtensionController to be set and we don&apos;t want to assume the default configuration.
(-[WKWebViewConfiguration _setWebExtensionController:]): Use nil instead of nullptr.
(-[WKWebViewConfiguration _weakWebExtensionController]): Ditto.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController init]): Use default configuration.
(-[_WKWebExtensionController initWithConfiguration:]): Added.
(-[_WKWebExtensionController configuration]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.mm: Added.
(+[_WKWebExtensionControllerConfiguration supportsSecureCoding]):
(+[_WKWebExtensionControllerConfiguration defaultConfiguration]):
(+[_WKWebExtensionControllerConfiguration nonPersistentConfiguration]):
(+[_WKWebExtensionControllerConfiguration configurationWithIdentifier:]):
(-[_WKWebExtensionControllerConfiguration encodeWithCoder:]):
(-[_WKWebExtensionControllerConfiguration initWithCoder:]):
(-[_WKWebExtensionControllerConfiguration copyWithZone:]):
(-[_WKWebExtensionControllerConfiguration dealloc]):
(-[_WKWebExtensionControllerConfiguration isEqual:]):
(-[_WKWebExtensionControllerConfiguration debugDescription]):
(-[_WKWebExtensionControllerConfiguration identifier]):
(-[_WKWebExtensionControllerConfiguration isPersistent]):
(-[_WKWebExtensionControllerConfiguration _apiObject]):
(-[_WKWebExtensionControllerConfiguration _webExtensionControllerConfiguration]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfigurationInternal.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfigurationPrivate.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::WebExtensionContext):
(WebKit::WebExtensionContext::load): Added a storageDirectory param. Call readStateFromStorage().
(WebKit::WebExtensionContext::unload): Call writeStateToStorage().
(WebKit::WebExtensionContext::stateFilePath const): Added.
(WebKit::WebExtensionContext::currentState const): Added.
(WebKit::WebExtensionContext::readStateFromStorage): Added.
(WebKit::WebExtensionContext::writeStateToStorage const): Added.
(WebKit::WebExtensionContext::setBaseURL): Use makeString() instead of StringBuilder.
(WebKit::WebExtensionContext::setUniqueIdentifier): Don&apos;t use baseURL&apos;s host, use a UUID for empty.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::storageDirectory const): Added.
(WebKit::WebExtensionController::load): Pass a storageDirectory to the loaded context.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfiguration.mm: Added.
(WebKit::WebExtensionControllerConfiguration::createStorageDirectoryPath):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::isPersistent const): Added.
(WebKit::WebExtensionContext::hasCustomUniqueIdentifier const): Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
(WebKit::WebExtensionController::WebExtensionController):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::create):
(WebKit::WebExtensionController::configuration const):
(WebKit::WebExtensionController::isPersistent const):
(WebKit::WebExtensionController::hasLoadedContexts const):
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.cpp: Added.
(WebKit::WebExtensionControllerConfiguration::WebExtensionControllerConfiguration):
(WebKit::WebExtensionControllerConfiguration::copy const):
(WebKit::WebExtensionControllerConfiguration::operator== const):
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.h: Added.
(WebKit::WebExtensionControllerConfiguration::createDefault):
(WebKit::WebExtensionControllerConfiguration::createNonPersistent):
(WebKit::WebExtensionControllerConfiguration::create):
(WebKit::WebExtensionControllerConfiguration::identifier const):
(WebKit::WebExtensionControllerConfiguration::isPersistent const):
(WebKit::WebExtensionControllerConfiguration::storageDirectory const):
(WebKit::WebExtensionControllerConfiguration::operator!= const):
(WebKit::WebExtensionControllerConfiguration::wrapper const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionController.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionControllerConfiguration.mm: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionManager initWithExtension:]):

Canonical link: <a href="https://commits.webkit.org/257310@main">https://commits.webkit.org/257310@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9fb471a5e324ec6c7fb3fd30355158abaa17c262

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98493 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7702 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31626 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107922 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168193 "Failed to checkout and rebase branch from PR 7077") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102441 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8218 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85094 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91042 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104575 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104162 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6235 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89791 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33228 "Found 2 new test failures: fast/text/text-edge-property-parsing.html, imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88054 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1643 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1564 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6489 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42101 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2523 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2941 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -253,11 +263,16 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
 
     URL m_baseURL;
     String m_uniqueIdentifier = UUID::createVersion4().toString();
+    bool m_customUniqueIdentifier { false };
```
This sounds like a variable name that would be the custom unique identifier itself, but since it's a bool it's only the presence of it? Maybe "has" could be in the name?

xeenon committed:
Add _WKWebExtensionControllerConfiguration to control extension storage.
https://bugs.webkit.org/show_bug.cgi?id=248665

Reviewed by Brian Weinstein.

_WKWebExtensionControllerConfiguration allows for default persistent storage, non-persistent,
and identifier based storage. With this, we can now store state for extension contexts and have
a directory structure to store extension data in the future (like local storage, etc).

Default storage is placed at: ~/Library/WebKit/<app-id>/WebExtensions/Default/<extension-id>
Non-persistent storage is placed at: ~/Library/WebKit/<app-id>/WebExtensions/<identifier>/<extension-id>
Non-persistent storage is not saved to disk.

Only extension contexts that have a custom uniqueIdentifier will be persistent as well, since the default
uniqueIdentifier is a random UUID, persisting those would cause state to be saved under random IDs that
can't be reused later.

* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Platform/cocoa/CocoaHelpers.h: Removed UUID helpers since UUID has Cocoa conversions now.
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm: Ditto.
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration copyWithZone:]): Adjust access to _webExtensionController now that it is RetainPtr.
(-[WKWebViewConfiguration _strongWebExtensionController]): Ditto.
(-[WKWebViewConfiguration _webExtensionController]): Don't use a LazyInitialized since we don't require a
webExtensionController to be set and we don't want to assume the default configuration.
(-[WKWebViewConfiguration _setWebExtensionController:]): Use nil instead of nullptr.
(-[WKWebViewConfiguration _weakWebExtensionController]): Ditto.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController init]): Use default configuration.
(-[_WKWebExtensionController initWithConfiguration:]): Added.
(-[_WKWebExtensionController configuration]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfiguration.mm: Added.
(+[_WKWebExtensionControllerConfiguration supportsSecureCoding]):
(+[_WKWebExtensionControllerConfiguration defaultConfiguration]):
(+[_WKWebExtensionControllerConfiguration nonPersistentConfiguration]):
(+[_WKWebExtensionControllerConfiguration configurationWithIdentifier:]):
(-[_WKWebExtensionControllerConfiguration encodeWithCoder:]):
(-[_WKWebExtensionControllerConfiguration initWithCoder:]):
(-[_WKWebExtensionControllerConfiguration copyWithZone:]):
(-[_WKWebExtensionControllerConfiguration dealloc]):
(-[_WKWebExtensionControllerConfiguration isEqual:]):
(-[_WKWebExtensionControllerConfiguration debugDescription]):
(-[_WKWebExtensionControllerConfiguration identifier]):
(-[_WKWebExtensionControllerConfiguration isPersistent]):
(-[_WKWebExtensionControllerConfiguration _apiObject]):
(-[_WKWebExtensionControllerConfiguration _webExtensionControllerConfiguration]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfigurationInternal.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerConfigurationPrivate.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::WebExtensionContext):
(WebKit::WebExtensionContext::load): Added a storageDirectory param. Call readStateFromStorage().
(WebKit::WebExtensionContext::unload): Call writeStateToStorage().
(WebKit::WebExtensionContext::stateFilePath const): Added.
(WebKit::WebExtensionContext::currentState const): Added.
(WebKit::WebExtensionContext::readStateFromStorage): Added.
(WebKit::WebExtensionContext::writeStateToStorage const): Added.
(WebKit::WebExtensionContext::setBaseURL): Use makeString() instead of StringBuilder.
(WebKit::WebExtensionContext::setUniqueIdentifier): Don't use baseURL's host, use a UUID for empty.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::storageDirectory const): Added.
(WebKit::WebExtensionController::load): Pass a storageDirectory to the loaded context.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerConfiguration.mm: Added.
(WebKit::WebExtensionControllerConfiguration::createStorageDirectoryPath):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::isPersistent const): Added.
(WebKit::WebExtensionContext::hasCustomUniqueIdentifier const): Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
(WebKit::WebExtensionController::WebExtensionController):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::create):
(WebKit::WebExtensionController::configuration const):
(WebKit::WebExtensionController::isPersistent const):
(WebKit::WebExtensionController::hasLoadedContexts const):
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.cpp: Added.
(WebKit::WebExtensionControllerConfiguration::WebExtensionControllerConfiguration):
(WebKit::WebExtensionControllerConfiguration::copy const):
(WebKit::WebExtensionControllerConfiguration::operator== const):
* Source/WebKit/UIProcess/Extensions/WebExtensionControllerConfiguration.h: Added.
(WebKit::WebExtensionControllerConfiguration::createDefault):
(WebKit::WebExtensionControllerConfiguration::createNonPersistent):
(WebKit::WebExtensionControllerConfiguration::create):
(WebKit::WebExtensionControllerConfiguration::identifier const):
(WebKit::WebExtensionControllerConfiguration::isPersistent const):
(WebKit::WebExtensionControllerConfiguration::storageDirectory const):
(WebKit::WebExtensionControllerConfiguration::operator!= const):
(WebKit::WebExtensionControllerConfiguration::wrapper const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionController.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionControllerConfiguration.mm: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionManager initWithExtension:]):

Canonical link: https://commits.webkit.org/257310@main
